docs/current_docs/manuals/developer/functions.mdx

```diff
@@ -338,6 +338,109 @@ The result will look like this:
 Hello, world
 ```
 
+## Contextual values
+
+Function arguments can define a default contextual value if no value is supplied for them. This default value will be load prior to the function call which makes it more cache efficient.
+
+:::note
+Contextual values are only available for arguments of type `Directory` and `File` for now.
+
+It's specially useful for static directory or file provided to the module. For example, a codebase source's code.
+:::
+
+The contextual value consist of a path to a directory or a file set through the `defaultPath` parameter metadata pointing to a location inside the module's context directory.
+- If the path is absolute, its root is the path to the git repository containting that module (called `context directory`), falling into the location of the `dagger.json` if no `.git` directory is found.
+- If the path is relative, its root is the path to `dagger.json` location (called `root directory`).
+
+Here's an example of Dagger Functions with a contextual arguments:
+- `repoFiles` has an argument `repo` of type `Directory` and a `defaultPath` parameter metadata set to `/` and returns the list of files in the given directory.
+- `moduleFiles` has an argument `module` of type `Directory` and a `defaultPath` parameter metadata set to `.` and returns the list of files in the given directory.
+- `readme` has an argument `readmeFile` of type `File` and a `defaultPath` parameter metadata set to `/README.md` and returns the contents of the given file.
+
+<Tabs groupId="language">
+
+<TabItem value="Go">
+
+For the example, we will use a repository with the following files:
+
+```shell
+.
+├── README.md
+├── file.txt
+└── my-module
+    ├── dagger.gen.go
+    ├── dagger.json
+    ├── go.mod
+    ├── go.sum
+    ├── internal
+    └── main.go
+```
+
+To define a contextual argument, you can add the `+defaultPath` pragma to the function parameter and set it to the path to default to.
+
+```go file=./snippets/functions/arguments-contextual/go/main.go
+```
+
+</TabItem>
+
+<TabItem value="Python">
+
+For the example, we will use a repository with the following files:
+
+```shell
+.
+├── README.md
+├── file.txt
+└── my-module
+    ├── LICENSE
+    ├── dagger.json
+    ├── package.json
+    ├── sdk
+    ├── src
+    ├── tsconfig.json
+    └── yarn.lock
+```
+
+To define a contextual argument, you can annotate the function parameter with the `DefaultPath` annotation and set it to the path to default to.
+
+```python file=./snippets/functions/arguments-contextual/python/main.py
+```
+
+</TabItem>
+
+<TabItem value="Typescript">
+
+For the example, we will use a repository with the following files:
```

Comment:
This file layout seems to be common, so could be outside the tabs.

Reply:
The files are not the same, because `Ts` has an `src` dir but not go for example.

Reply:
I see. But this is happening because we are specifically using a module dir in our example functions. If we change the functions to make them more generic, such as a generic read dir function and read file function (see my other comment further down) then you could have a simple dir structure like below which doesn't need to be language-specific.

```
README-main.md
a/
  README-a.md
  b/ 
    README-b.md
    c/
      README-c.md
```

Not a big deal tbh, just trying to optimize the example :)

Reply:
Yeah I totally understand and agree with you! I'll do this